### ISSUE TITLE
Inconsistencies in NOT EXISTS and EXISTS nodes in the algebra tree

### ISSUE THREAD

**Issue by GreenfishK** [RDFLib/rdflib#2131]
When traversing a SPARQL query tree using the `algebra.traverse` function I would expect to find a consistent view of any subtree that I am looking at during the traversel with the parent tree. However, when I arrive at the NOT EXISTS or EXISTS node the tree that I get when I view the node and it's child node.graph are not consistent anymore. node.graph resolves to something totally different. This leads to some issues when doing modifications on the algebra tree during the traversel. Below you can find a fully reproducible example:

Let's start with an example SPARQL query and translate it into an algebra tree
```
import rdflib.plugins.sparql.parser as parser
import rdflib.plugins.sparql.algebra as algebra
from rdflib.plugins.sparql.parserutils import CompValue

query = """
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

select ?p ?o {
	<http://dbpedia.org/resource/Donald_Trump> ?p ?o .
    filter not exists {
        <http://dbpedia.org/resource/Donald_Trump> rdf:type ?o
    }
}
"""

query_tree = parser.parseQuery(query)
query_algebra = algebra.translateQuery(query_tree)
algebra.pprintAlgebra(query_algebra)
print("\n\n")
```
This yields the output:
```
SelectQuery(
    p = Project(
        p = Filter(
            expr = Builtin_NOTEXISTS(
                graph = GroupGraphPatternSub(
                    part = [TriplesBlock_{'triples': [[rdflib.term.URIRef('http://dbpedia.org/resource/Donald_Trump'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), rdflib.term.Variable('o')]], '_vars': {rdflib.term.Variable('o')}}]
                    _vars = {rdflib.term.Variable('o')}
                    )
                _vars = {rdflib.term.Variable('o')}
                )
            p = BGP(
                triples = [(rdflib.term.URIRef('http://dbpedia.org/resource/Donald_Trump'), rdflib.term.Variable('p'), rdflib.term.Variable('o'))]
                _vars = {rdflib.term.Variable('o'), rdflib.term.Variable('p')}
                )
            _vars = {rdflib.term.Variable('o'), rdflib.term.Variable('p')}
            )
        PV = [rdflib.term.Variable('p'), rdflib.term.Variable('o')]
        _vars = {rdflib.term.Variable('o'), rdflib.term.Variable('p')}
        )
    datasetClause = None
    PV = [rdflib.term.Variable('p'), rdflib.term.Variable('o')]
    _vars = {rdflib.term.Variable('o'), rdflib.term.Variable('p')}
    )

```

Now, let's traverse the tree and stop at a node called NOTEXISTS. Once we are there, we print the node, i.e. the subtree and the subtree of node.graph. A NOTEXISTS node seems to always have such a child.

```

def check_not_exists(node: CompValue):
    if isinstance(node, CompValue):
        if node.name.endswith("NOTEXISTS"):
            print(node, end="\n\n")
            print(node.graph)

node_algebra = algebra.traverse(query_algebra.algebra, visitPre=check_not_exists)

```

The output I get is:
```
Builtin_NOTEXISTS_Builtin_NOTEXISTS_{'graph': GroupGraphPatternSub_{'part': [TriplesBlock_{'triples': [[rdflib.term.URIRef('http://dbpedia.org/resource/Donald_Trump'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), rdflib.term.Variable('o')]], '_vars': {rdflib.term.Variable('o')}}], '_vars': {rdflib.term.Variable('o')}}, '_vars': {rdflib.term.Variable('o')}}

Join_Join_{'p1': BGP_{'triples': []}, 'p2': BGP_{'triples': [(rdflib.term.URIRef('http://dbpedia.org/resource/Donald_Trump'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), rdflib.term.Variable('o'))]}}
```

The Output I would expect to get from `print(node.graph)` is something like:
```
GroupGraphPatternSub_{'part': [TriplesBlock_{'triples': [[rdflib.term.URIRef('http://dbpedia.org/resource/Donald_Trump'), rdflib.term.URIRef('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), rdflib.term.Variable('o')]], '_vars': {rdflib.term.Variable('o')}}], '_vars': {rdflib.term.Variable('o')}}, '_vars': {rdflib.term.Variable('o')}
```
So basically I would expect to see a subset of print(node). This is usually the behavior with other nodes, but here, node.graph resolves to something different. Is this behavior expected?
